controllers/storageautoscaler/storageautoscaler_controller.go

```diff
@@ -0,0 +1,255 @@
+
+		case <-time.After(r.RequeueTime):
+			r.reconcileMetrics()
+		}
+	}
+}
+
+func (r *StorageAutoscalerReconciler) reconcileMetrics() {
+	r.Log.Info("scraping metrics")
+	metrics, err := r.scrapeMetrics()
+	if err != nil {
+		r.Log.Error(err, "failed to scrape metrics")
+		return
+	}
+	r.Log.Info("scraped metrics", "metrics", metrics)
+
+	r.Log.Info("update sync map")
+	r.updateSyncMap(metrics)
+	r.Log.Info("sync map updated")
+
+	r.Log.Info("get resource objects from kube api that need scaling")
+	objectList, err := r.getStorageAutoScalerObjects()
+	if err != nil {
+		r.Log.Error(err, "failed to get resource objects")
+		return
+	}
+	r.Log.Info("resource objects fetched")
+
+	r.Log.Info("send a signal to reconcile using generic event")
+	r.sendGenericEvent(objectList)
+	r.Log.Info("generic event sent")
+}
+
+func (r *StorageAutoscalerReconciler) scrapeMetrics() (model.Vector, error) {
+	prometheusURL := "https://prometheus-k8s.openshift-monitoring.svc.cluster.local:9091"
+	if isROSAHCP, err := platform.IsPlatformROSAHCP(); err != nil {
+		r.Log.Error(err, "Failed to determine if ROSA HCP cluster")
+		return nil, err
+	} else if isROSAHCP {
+		prometheusURL = fmt.Sprintf("https://prometheus.%s.svc.cluster.local:9339", r.OperatorNamespace)
+	}
+
+	caCertPath := "/var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt"
+	tokenPath := "/var/run/secrets/kubernetes.io/serviceaccount/token"
+
+	// create auth roundtripper
+	tlsConfig, err := prometheusconfig.NewTLSConfig(&prometheusconfig.TLSConfig{
+		CAFile: caCertPath,
+	})
+	if err != nil {
+		r.Log.Error(err, "failed to create tls config")
+		return nil, err
+	}
+	settings := prometheusconfig.TLSRoundTripperSettings{}
+	newRT := func(cfg *tls.Config) (http.RoundTripper, error) {
+		return &http.Transport{TLSClientConfig: cfg}, nil
+	}
+	tlsRoundTripper, err := prometheusconfig.NewTLSRoundTripperWithContext(r.ctx, tlsConfig, settings, newRT)
+	if err != nil {
+		r.Log.Error(err, "failed to create tls round tripper")
+		return nil, err
+	}
+	roundTripper := prometheusconfig.NewAuthorizationCredentialsRoundTripper("Bearer", prometheusconfig.NewFileSecret(tokenPath), tlsRoundTripper)
+	// create prometheus client
+	prometheusClient, err := prometheusapi.NewClient(prometheusapi.Config{
+		Address:      prometheusURL,
+		RoundTripper: roundTripper,
+	})
+	if err != nil {
+		r.Log.Error(err, "failed to create prometheus client")
+		return nil, err
+	}
+	// scrape the metrics
+	scraper := prometheusv1.NewAPI(prometheusClient)
+	osdUsageQuery := "(ceph_osd_metadata * on (ceph_daemon, namespace, managedBy) group_right(device_class,hostname) (ceph_osd_stat_bytes_used / ceph_osd_stat_bytes))"
+	result, warnings, err := scraper.Query(context.Background(), osdUsageQuery, time.Now(), prometheusv1.WithTimeout(10*time.Second))
+	if err != nil {
+		r.Log.Error(err, "failed to query prometheus")
+		return nil, err
+	}
+
+	if len(warnings) > 0 {
+		r.Log.Info("prometheus query warnings", "warnings", warnings)
```

Comment:
Should we surface warnings as info or warnings?

```suggestion
		r.Log.Warning("prometheus query warnings", "warnings", warnings)
```

Reply:
I don't see if there is nay warning method with the current logger, @iamniting are you aware

Reply:
We use klog which doesn't have Warning level logs. Kubernetes best practices for structured logging suggests that logs are either critical enough to be Error level or not critical at all, so Info level.